File: built-in plugins dsp/mb-bass-808.cpp
/**
 * MB 808 Bass
 * Category : instrument
 * Type     : bass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic 808 kick bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BASS_808_H
#define MB_BASS_808_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBass808 : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bass-808";
    static constexpr const char* PLUGIN_NAME    = "MB 808 Bass";
    static constexpr const char* PLUGIN_TYPE    = "bass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float slide = 0.3f;  // range [0, 1]
    float volume = 0.9f;  // range [0, 1]
    };

    MbBass808() = default;
    ~MbBass808() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.slide = std::clamp(params.slide, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB 808 Bass
        return input;
    }
};

#endif // MB_BASS_808_H

File: built-in plugins dsp/mb-bass-808.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbBass808Plugin: PluginDefinition = {
    id: 'mb-bass-808', slug: 'mb-bass-808', name: 'MB 808 Bass', category: 'instrument', type: 'bass', version: '1.0.0',
    description: 'Classic 808 kick bass', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'sine', detune: 0, gain: 1.0 }],
    envelope: { attack: 0.001, decay: 0.8, sustain: 0.0, release: 0.5 },
    parameters: [
      { id: 'slide', name: 'Slide', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.9, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { slide: 0.3, volume: 0.9 },
  };

export default MbBass808Plugin;
